projects/newprofile_ping_nightly_validation.kp/knowledge.md

```diff
@@ -0,0 +1,518 @@
+
+### Cross-check the `new-profile` and `main` pings 
+
+
+```python
+main_pings = Dataset.from_source("telemetry") \
+                    .where(docType='main') \
+                    .where(appUpdateChannel="nightly") \
+                    .where(submissionDate=lambda x: "20170531" <= x < "20170607") \
+                    .where(appBuildId=lambda x: "20170531" <= x < "20170607") \
+                    .records(sc, sample=1.0)
+```
+    fetching 23348.80037MB in 8027 files...
+
+
+
+```python
+main_subset = get_pings_properties(main_pings, ["id",
+                                                "meta/creationTimestamp",
+                                                "meta/Date",
+                                                "meta/Timestamp",
+                                                "meta/X-PingSender-Version",
+                                                "clientId",
+                                                "environment/profile/creationDate",
+                                                "payload/info/reason",
+                                                "payload/info/sessionLength",
+                                                "payload/info/subsessionLength",
+                                                "payload/info/profileSubsessionCounter",
+                                                "payload/info/previousSessionId"])
+```
+Dedupe by document id and restrict the `main` ping data to the pings from the misbehaving and well behaving clients.
+
+
+```python
+well_behaving_clients =\
+    set(subset.filter(lambda p: p.get("clientId") not in misbehaving_clients).map(lambda p: p.get("clientId")).collect())
+
+all_clients = misbehaving_clients + list(well_behaving_clients)
+```
+
+```python
+main_deduped = dedupe(main_subset.filter(lambda p: p.get("clientId") in all_clients), "id")
+main_deduped_count = main_deduped.count()
+```
+Try to pair each `new-profile` ping with reason `shutdown` to the very first `main` ping with reason `shutdown` received from that client, to make sure that the former were sent at the right time.
+
+
+```python
+first_main = main_deduped.filter(lambda p:\
+                                    p.get("payload/info/previousSessionId") == None and\
+                                    p.get("payload/info/reason") == "shutdown")
+```
+
+```python
+newping_shutdown = deduped_docid.filter(lambda p: p.get("payload/reason") == "shutdown")
+```
+
+```python
+newprofile_plus_main = first_main.union(newping_shutdown)
+sorted_per_client = newprofile_plus_main.map(lambda p: (p["clientId"], [(p, p["meta/creationTimestamp"])]))\
+                                        .reduceByKey(lambda a, b: sorted(a + b, key=lambda k: k[1]))\
```

Comment:
The 'new-profile' should always be generated before the 'main', at shutdown. See [here](http://searchfox.org/mozilla-central/rev/d840ebd5858a61dbc1622487c1fab74ecf235e03/toolkit/components/telemetry/TelemetryController.jsm#796,806) where we wait for the 'new-profile' to be generated before moving on to shut down TelemetrySession (which generates the 'main' ping). Although this is how it happens, we're not making any strict promise on that.